docs/source/model/models-from-code.rst

```diff
@@ -0,0 +1,491 @@
+        operations, and then use this function within an ``ArithmeticModel`` custom ``PythonModel`` that we will define in a separate script. 
+        This model will be logged with MLflow, allowing us to perform predictions using the stored model.
+
+        This tutorial will show you how to:
+
+        - Create multiple Python files from within a Jupyter notebook.
+        - Log a custom model with MLflow that relies on external code defined in another file.
+        - Use the ``code_paths`` feature to include additional scripts when logging the model, ensuring that all dependencies are available when the model is loaded for inference.
+
+        In the first step, we define our ``add`` function in a file named ``calculator.py``, including the magic ``%%writefile`` command if we're running in a notebook cell:
+
+        
+        .. code-block:: python
+
+            # If running in a Jupyter or Databricks notebook cell, uncomment the following line:
+            # %%writefile "./calculator.py"
+
+
+            def add(x, y):
+                return x + y
+
+        Next, we create a new file, ``math_model.py``, which contains the ``ArithmeticModel`` class. This script will be responsible for importing the ``add`` function from our external script, defining our model, 
+        performing predictions, and validating the input data types. The predict method will leverage the ``add`` function to perform the addition of two numbers provided as input.
+
+        The following code block writes the ``ArithmeticModel`` class definition to ``math_model.py``:
+
+        .. code-block:: python
+
+            # If running in a Jupyter or Databricks notebook cell, uncomment the following line:
+            # %%writefile "./math_model.py"
+
+            from typing import Dict, Any, Union
+            from mlflow.pyfunc import PythonModel
+            from mlflow.models import set_model
+
+            from calculator import add
+
+
+            class ArithmeticModel(PythonModel):
+                def __init__(self):
+                    self.types = (int, float, complex)
+
+                def predict(
+                    self, context, model_input: Dict[str, Any], params=None
+                ) -> Union[int, float, complex]:
+                    try:
+                        a = model_input["a"]
+                        b = model_input["b"]
+                    except KeyError as e:
+                        raise KeyError(f"Missing required input: {e.args[0]}") from e
+
+                    if not isinstance(a, self.types) or not isinstance(b, self.types):
+                        raise ValueError(
+                            f"Input types must be one of {self.types}, but received: {type(a)}, {type(b)}"
+                        )
+
+                    # Directly use a dependency that is defined within code_paths in the logged model
+                    return add(a, b)
```

Comment:
```suggestion
            class AddModel(PythonModel):
                def __init__(self):
                    self.types = (int, float, complex)

                def predict(
                    self, context, model_input: Dict[str, Any], params=None
                ) -> Union[int, float, complex]:
                    return add(model_input["a"], model_input["b"])
```

can we further simplify the example?

Reply:
yeah good call. Stripped it down to only essentials.